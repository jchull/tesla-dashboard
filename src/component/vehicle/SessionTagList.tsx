import React, {useEffect, useState} from 'react';
import {TagList} from '../common/TagList';
import {queryService} from '@service/Services';
import {SessionTagListProps} from '../../type/state';

export const SessionTagList: React.FC<SessionTagListProps> = props => {

  function tagsToDisplay(tags: string[]): string[] {
    return tags.map(tag => tag.replace('_', ' '));
  }

  async function addListener(tag: string): Promise<any> {
    const newTags = await queryService.addTag(props.vehicleId, props.sessionId, tag);
    setTags(tagsToDisplay(newTags));
  }

  async function removeListener(tag: string): Promise<any> {
    const newTags = await queryService.removeTag(props.vehicleId, props.sessionId, tag);
    setTags(tagsToDisplay(newTags));
  }


  const [tags, setTags] = useState(props.tags || []);

  useEffect(() => {
    setTags(tagsToDisplay(props.tags));
  }, [props.sessionId]);


  return (
      <TagList tags={tags}
               removeListener={removeListener}
               addListener={addListener}/>
  );
};

