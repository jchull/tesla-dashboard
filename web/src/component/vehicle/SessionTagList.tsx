import { TagList } from '../tags/TagList';
import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { AppState } from '@store/store';
import { createSelector } from '@reduxjs/toolkit';
import {
  addSessionTagAction,
  removeSessionTagAction
} from '@component/session/actions';

function tagsToDisplay(tags: string[]): string[] {
  return tags.map((tag) => tag.replace('_', ' '));
}

export const SessionTagList: React.FC = () => {
  const dispatch = useDispatch();
  const sessionIdSelector = (store: AppState) =>
    store.session.selectedSessionId;
  const sessionsSelector = (store: AppState) => store.session.sessions;
  const tagsSelector = createSelector(
    [sessionsSelector, sessionIdSelector],
    (sessions, selectedSessionId) => {
      const session = sessions.find(
        (session) => session._id === selectedSessionId
      );
      return tagsToDisplay(session && session.tags ? session.tags : []);
    }
  );

  const tags = useSelector(tagsSelector);
  const sessionId = useSelector(sessionIdSelector);

  async function addListener(tag: string): Promise<void> {
    if (sessionId && tag && tag.length) {
      dispatch(addSessionTagAction(sessionId, tag));
    }
  }

  async function removeListener(tag: string): Promise<void> {
    if (sessionId && tag && tag.length) {
      dispatch(removeSessionTagAction(sessionId, tag));
    }
  }

  return (
    <TagList
      tags={tags}
      removeListener={removeListener}
      addListener={addListener}
    />
  );
};
