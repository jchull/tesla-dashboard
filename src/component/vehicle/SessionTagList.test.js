import { act, render } from '@testing-library/react';
import pretty          from 'pretty';
import React           from 'react';


import { SessionTagList } from './SessionTagList';

let wrapper;

beforeEach(() => {
});

afterEach(() => {
  if(wrapper) {
    if(wrapper.unmount) {
      wrapper.unmount();
    }
    if(wrapper.container) {
      wrapper.container = null;
    }
  }
});


it('renders empty tag list', () => {
  act(() => {
    wrapper = render(<SessionTagList tags={[]}
                                     sessionId="xyz1"
                                     vehicleId="abc1"/>);
  });
  expect(pretty(wrapper.container.innerHTML))
    .toMatchSnapshot();
});

