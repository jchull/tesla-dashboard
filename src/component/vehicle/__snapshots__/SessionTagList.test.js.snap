// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders empty tag list 1`] = `
"<div class=\\"tag-list\\">
  <div class=\\"tag-input\\"><button class=\\"add\\" aria-label=\\"Show new tag entry\\"><i class=\\"material-icons\\">add</i></button><input class=\\"hide\\" aria-label=\\"Enter new tag\\"></div>
</div>"
`;
